docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+---
+title: ClusterClass and managed topologies
+authors:
+  - "@srm09"
+  - "@vincepri"
+  - "@fabriziopandini"
+reviewers:
+  - "@vincepri"
+  - "@fabriziopandini"
+creation-date: 2021-05-26
+status: provisional
+replaces:
+  - [Proposal Google Doc](https://docs.google.com/document/d/1lwxgBK3Q7zmNkOSFqzTGmrSys_vinkwubwgoyqSRAbI/edit#)
+---
+
+# ClusterClass and Managed Topologies
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+      - [ClusterClass](#clusterclass)
+      - [Managed Cluster topology](#managed-cluster-topology)
+      - [Worker class](#worker-class)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+      - [Goals](#goals)
+      - [Prospective future Work](#prospective-future-work)
+  - [Proposal](#proposal)
+      - [User Stories](#user-stories)
+        - [Story 1 - Use ClusterClass to easily stamp clusters](#story-1---use-clusterclass-to-easily-stamp-clusters)
+        - [Story 2 - Easier UX for kubernetes version upgrades](#story-2---easier-ux-for-kubernetes-version-upgrades)
+        - [Story 3 - Easier UX for scaling workers nodes](#story-3---easier-ux-for-scaling-workers-nodes)
+      - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+        - [New API types](#new-api-types)
+            - [ClusterClass](#clusterclass-1)
+        - [Modification to existing API Types](#modification-to-existing-api-types)
+            - [Cluster](#cluster)
+        - [Validations](#validations)
+            - [ClusterClass](#clusterclass-2)
+            - [Cluster](#cluster-1)
+        - [Behaviors](#behaviors)
+            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
+            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
+        - [Provider implementation](#provider-implementation)
+            - [For infrastructure providers](#for-infrastructure-providers)
+            - [For Control plane providers](#for-control-plane-providers)
+      - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+      - [Test Plan [optional]](#test-plan-optional)
+      - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+### ClusterClass
+A collection of templates that define a topology (control plane and machine deployments) to be used in one or more Managed Clusters.
+
+### Managed Cluster topology
+A Managed Cluster is a Cluster that provides a single control point to manage it's own topology; the managed cluster topology is defined by a ClusterClass.
+
+### Worker class
```

Comment:
So is a `clusterClass` composed of 0..N `workerClass`? if so probably worth call that out in this glossary.

Reply:
It is not just a collection of worker classes but also defines the control plane template. I can add something similar to it though.